modules/apps/powwow/powwow-portlet/docroot/WEB-INF/src/com/liferay/powwow/util/PowwowUtil.java

```diff
@@ -336,6 +354,85 @@ public static PowwowSubscriptionSender getPowwowSubscriptionSender(
 		return powwowSubscriptionSender;
 	}
 
+	public static String getRecurrenceSummary(Recurrence recurrence, ServiceContext serviceContext) {
+
+		String recurrenceSummary = StringPool.BLANK;
+
+		Locale locale = serviceContext.getLocale();
+
+		StringBuilder sb = new StringBuilder();
+
+		Map<String, Object> valuesMap = new HashMap<>();
+
+		String position = StringPool.BLANK;
+		String weekDay = StringPool.BLANK;
+		String weekDays = StringPool.BLANK;
+
+		if (recurrence.getInterval() == 1) {
+			sb.append(StringUtils.capitalize(recurrence.getFrequency().getValue()));
+		}
+		else {
+			sb.append(LanguageUtil.get(locale, "every") + " ${interval} ${frequencyLabel}");
+
+			valuesMap.put("interval", recurrence.getInterval());
+			valuesMap.put("frequencyLabel", _frequencyMapLabel.get(recurrence.getFrequency()));
+		}
+
+		if (Validator.isNotNull(recurrence.getPositionalWeekday()) && recurrence.getFrequency() == Frequency.MONTHLY) {
+			PositionalWeekday positionalWeekday = recurrence.getPositionalWeekday();
+
+			sb.append(STR_SPACE + LanguageUtil.get(locale, "on") + " ${position} ${weekDay}");
+
+			position = _positionMapLabel.get(positionalWeekday.getPosition());
+			weekDay = _weekDayMapLabel.get(positionalWeekday.getWeekday());
+
+			valuesMap.put("position", position);
+			valuesMap.put("weekDay", weekDay);
+		}
+		else if (recurrence.getFrequency() == Frequency.WEEKLY && recurrence.getWeekdays().size() > 0) {
+			sb.append(STR_SPACE + LanguageUtil.get(locale, "on") + " ${weekDays}");
+
+			List<String> weekdayList = new ArrayList<>();
+
+			for (Weekday day : recurrence.getWeekdays()) {
+				weekdayList.add(_weekDayMapLabel.get(day));
+			}
+
+			weekDays = weekdayList.stream().collect(Collectors.joining(","));
+
+			valuesMap.put("weekDays", weekDays);
+		}
+
+		if (recurrence.getCount() > 0) {
+			sb.append(", ${count} " + LanguageUtil.get(locale, "times"));
+
+			valuesMap.put("count", recurrence.getCount());
+		}
+		else if (Validator.isNotNull(recurrence.getUntilJCalendar())) {
+			Calendar untilDateJCalendar = recurrence.getUntilJCalendar();
+			LocalDate untilDate = LocalDate.of(untilDateJCalendar.get(Calendar.YEAR),
+				untilDateJCalendar.get(Calendar.MONTH), untilDateJCalendar.get(Calendar.DATE));
+
+			sb.append(", ");
+			sb.append(LanguageUtil.get(locale, "until"));
+			sb.append(STR_SPACE);
+			sb.append("${utilMonthLabel}");
+			sb.append(STR_SPACE);
+			sb.append("${utilDay}");
+			sb.append(STR_SPACE + ", ");
+			sb.append("${utilYear}");
+
+			valuesMap.put("utilMonthLabel", _monthMapLabel.get(untilDate.getMonthValue()));
+			valuesMap.put("utilDay",untilDate.getDayOfMonth());
+			valuesMap.put("utilYear", untilDate.getYear());
```

Comment:
eventually, with our need in this logic, i'm thinking that we can use a formatter, with the help of `Locale` and `TimeZone` then we can produce the string by a pattern. Do you think so?

Reply:
for this text, i think we can continue with current way, it will keep consistent with what the JS functions provide.